src/nupic/ntypes/Sdr.hpp

```diff
@@ -748,32 +762,87 @@ class SparseDistributedRepresentation : public Serializable
         string marker;
         UInt version;
         inStream >> marker >> version;
-        NTA_ASSERT( marker == "SDR" );
-        NTA_ASSERT( version == SERIALIZE_VERSION );
+        NTA_CHECK( marker == "SDR" );
+        NTA_CHECK( version == SERIALIZE_VERSION );
 
         // Read the dimensions.
         readVector( dimensions_ );
 
-        // Read the data.
-        clear();
-        readVector( flatSparse );
-        flatSparse_valid = true;
-
-        // Consume the end marker.
-        inStream >> marker;
-        NTA_ASSERT( marker == "~SDR" );
-
         // Initialize the SDR.
         // Calculate the SDR's size.
         size_ = 1;
         for(UInt dim : dimensions)
             size_ *= dim;
         // Initialize sparse tuple.
         sparse.assign( dimensions.size(), {} );
+
+        // Read the data.
+        readVector( flatSparse );
+        setFlatSparseInplace();
+
+        // Consume the end marker.
+        inStream >> marker;
+        NTA_CHECK( marker == "~SDR" );
     };
+
+    void addCallback(SDR_callback_t callback) {
+        callbacks.push_back( callback );
+    }
 };
 
 typedef SparseDistributedRepresentation SDR;
 
+/**
+ * TODO: Docs
+ */
+class SDR_Proxy : public SDR
+{
+public:
+    /**
+     * TODO: Docs
+     */
+    SDR_Proxy(SDR &sdr)
+        : SDR_Proxy(sdr, sdr.dimensions)
+        {}
+
+    SDR_Proxy(SDR &sdr, const vector<UInt> &dimensions)
+        : SDR( dimensions ) {
+        clear();
+        parent = &sdr;
+        NTA_CHECK( size == parent->size ) << "SDR Proxy must have same size as given SDR.";
+        parent->addCallback( [&] () {
+            clear();
+            do_callbacks();
+        });
+    };
+
+    SDR_dense_t& getDense() override
+        { return parent->getDense(); }
+
+    SDR_flatSparse_t& getFlatSparse() override
+        { return parent->getFlatSparse(); }
+
+    // Don't override getSparse().  It will call either getDense() or
+    // getFlatSparse() to get its data.
+
+private:
+    /**
+     * This SDR shall always have the same value as the parent SDR.
+     */
+    SDR *parent;
```

Comment:
use `shared_ptr<SDR> parent; `

Reply:
even const shared_ptr<SDR> as the class can never change its parent after init.

Reply:
Actually it can change its parent class after init.  It could request one of the data formats which it delegates to the parent (that is to say: dense and flat-sparse) and those could need to be recalculated, and of course the parent would want to save them in case they're used multiple times.  All of the getter methods are non-constant.

Reply:
I don't think we want unique_ptr either since that will try to free the SDR when the Proxy goes out of scope.

Reply:
It is possible to give unique_ptr a distructor function and will not free the pointer.  But that would be the same as passing a raw pointer around except that the unique_ptr would prevent anything in which it is contained from being copied.  Perhaps a raw pointer is sufficient.

Reply:
ok, good example. 
On the other hand, with the current raw pointer: 
- Proxy goes out of scope, it was last user of SDR -> mem leak (we don't check & delete parent in destructor) 
- SDR deleted while Proxy still alive -> our parent's memory is garbage

So, something like this? 
- Proxy has no constructor
- SDR has `unique_ptr<Proxy> getReadOnlyView(newDimensions)` that returns Proxy, if SDR is deleted, so is Proxy

Reply:
getReadOnlyView() would be `make_unique<SDR>((Proxy)&this); `

Reply:
I would really rather not do this. It adds complexity to the SDR and Proxy.  I dont see this as a big issue.  

I can add a deconstructor to both the SDR and Proxy which will cause an immediate & obvious error if used after deconstruction, but that is a mitigation not a true fix since the structure might be overwritten with arbitrary data.

Reply:
I don't think you want to use unique_ptr either.  If there can be more than one class that can be owner of the pointer then you need a shared_ptr.  A unique_ptr would just make sure there could not be more than one owner (cannot copy).  The shared_ptr will keep the pointer alive until both SDR and SDR_Proxy go out of scope.  And you can have more than one Proxy without getting into problems figuring out when it is safe to delete the pointer.  The down side is that there is a slight overhead for incrementing reference counts when Proxies are created/removed.  But that overhead would be very small.

Reply:
Summary of latest commit:

SDR_Proxy can now be safely deleted.
SDR_Proxy now throws error if used after parent SDR is deleted.
Whoever allocates SDR & SDR_Proxy is responsible for freeing them.